panic during shutdown

```
panic: DB Closed

goroutine 192 [running]:
github.com/dgraph-io/badger/v4.(*Txn).NewIterator(0xc014eb6580, {0x0, 0x0, 0x1, 0x0, 0x0, 0x0, {0xc00ca14e7e, 0x2, 0x2}, ...})
	/home/chris/go/pkg/mod/github.com/dgraph-io/badger/v4@v4.9.0/iterator.go:464 +0x7f4
github.com/blinklabs-io/dingo/database/plugin/blob/badger.(*BlobStoreBadger).NewIterator(0xc000482000, {0x4baecd0, 0xc00ac75920}, {{0xc00ca14e7e?, 0x0?, 0x0?}, 0x0?})
	/home/chris/src/blinklabs.io/dingo/database/plugin/blob/badger/database.go:370 +0x31c
github.com/blinklabs-io/dingo/database.BlocksRecentTxn(0xc01547c640, 0x64)
	/home/chris/src/blinklabs.io/dingo/database/block.go:250 +0x18b
github.com/blinklabs-io/dingo/database.BlocksRecent.func1(0xc01547c640)
	/home/chris/src/blinklabs.io/dingo/database/block.go:226 +0x45
github.com/blinklabs-io/dingo/database.(*Txn).Do(0xc01547c640, 0xc00081dad0)
	/home/chris/src/blinklabs.io/dingo/database/txn.go:91 +0x4c
github.com/blinklabs-io/dingo/database.BlocksRecent(0xc00ce1cd20, 0x64)
	/home/chris/src/blinklabs.io/dingo/database/block.go:224 +0xaf
github.com/blinklabs-io/dingo/ledger.(*LedgerState).RecentChainPoints(0xc00048f8c8, 0x64)
	/home/chris/src/blinklabs.io/dingo/ledger/state.go:1365 +0x6b
github.com/blinklabs-io/dingo/ouroboros.(*Ouroboros).chainsyncClientStart(0xc00ce1a360, {{0x4bad6c8?, 0xc015534b40?}, {0x4bad6c8?, 0xc015534b70?}})
	/home/chris/src/blinklabs.io/dingo/ouroboros/chainsync.go:64 +0x106
github.com/blinklabs-io/dingo/ouroboros.(*Ouroboros).HandleOutboundConnEvent(0xc00ce1a360, {{0xc2541589a05717fe, 0x99609b773e4, 0x6755040}, {0x403cc80, 0xc00057e5a0}, {0x44c3600, 0x15}})
	/home/chris/src/blinklabs.io/dingo/ouroboros/ouroboros.go:295 +0x1df
github.com/blinklabs-io/dingo/event.(*EventBus).SubscribeFunc.func1(...)
	/home/chris/src/blinklabs.io/dingo/event/event.go:169
created by github.com/blinklabs-io/dingo/event.(*EventBus).SubscribeFunc in goroutine 71
	/home/chris/src/blinklabs.io/dingo/event/event.go:163 +0x126

Process finished with the exit code 2
```

--- a/database/plugin/blob/badger/database.go
+++ b/database/plugin/blob/badger/database.go
@@ -527,6 +527,7 @@ func (d *BlobStoreBadger) NewIterator(
 	txn types.Txn,
 	opts types.BlobIteratorOptions,
 ) types.BlobIterator {
+	// ) (iter types.BlobIterator) {
 	badgerTxn, err := d.validateTxn(txn)
 	if err != nil {
 		return &errorIterator{err: err}
@@ -535,6 +536,18 @@ func (d *BlobStoreBadger) NewIterator(
 		Prefix:  opts.Prefix,
 		Reverse: opts.Reverse,
 	}
+	// defer func() {
+	// 	if r := recover(); r != nil {
+	// 		err, ok := r.(error)
+	// 		if ok && errors.Is(err, badger.ErrDBClosed) {
+	// 			iter = &errorIterator{
+	// 				err: fmt.Errorf("%w: %w", types.ErrBlobStoreUnavailable, err),
+	// 			}
+	// 			return
+	// 		}
+	// 		panic(r)
+	// 	}
+	// }()
 	return &badgerIterator{iter: badgerTxn.tx.NewIterator(iterOpts)}
 }
 

--- a/database/plugin/blob/badger/database.go
+++ b/database/plugin/blob/badger/database.go
@@ -526,8 +526,7 @@ func (d *BlobStoreBadger) Delete(txn types.Txn, key []byte) error {
 func (d *BlobStoreBadger) NewIterator(
 	txn types.Txn,
 	opts types.BlobIteratorOptions,
-) types.BlobIterator {
-	// ) (iter types.BlobIterator) {
+) (iter types.BlobIterator) {
 	badgerTxn, err := d.validateTxn(txn)
 	if err != nil {
 		return &errorIterator{err: err}
@@ -536,18 +535,18 @@ func (d *BlobStoreBadger) NewIterator(
 		Prefix:  opts.Prefix,
 		Reverse: opts.Reverse,
 	}
-	// defer func() {
-	// 	if r := recover(); r != nil {
-	// 		err, ok := r.(error)
-	// 		if ok && errors.Is(err, badger.ErrDBClosed) {
-	// 			iter = &errorIterator{
-	// 				err: fmt.Errorf("%w: %w", types.ErrBlobStoreUnavailable, err),
-	// 			}
-	// 			return
-	// 		}
-	// 		panic(r)
-	// 	}
-	// }()
+	defer func() {
+		if r := recover(); r != nil {
+			err, ok := r.(error)
+			if ok && errors.Is(err, badger.ErrDBClosed) {
+				iter = &errorIterator{
+					err: fmt.Errorf("%w: %w", types.ErrBlobStoreUnavailable, err),
+				}
+				return
+			}
+			panic(r)
+		}
+	}()
 	return &badgerIterator{iter: badgerTxn.tx.NewIterator(iterOpts)}
 }
 

--- a/database/plugin/blob/badger/database_test.go
+++ b/database/plugin/blob/badger/database_test.go
@@ -20,6 +20,7 @@ import (
 
 	"github.com/blinklabs-io/dingo/database/types"
 	"github.com/blinklabs-io/gouroboros/cbor"
+	"github.com/stretchr/testify/require"
 )
 
 func TestMarshalBlockMetadataRoundTrip(t *testing.T) {
@@ -85,4 +86,31 @@ func TestUnmarshalBlockMetadataLegacyCbor(t *testing.T) {
 	if !bytes.Equal(decoded.PrevHash, expected.PrevHash) {
 		t.Fatalf("expected PrevHash %x, got %x", expected.PrevHash, decoded.PrevHash)
 	}
+}
+
+// It reproduces the shutdown race where a transaction created before BlobStoreBadger.Close()
+// is later used to create an iterator. Badger panics with "DB Closed" in
+// that case; the blob store should return an error iterator instead.
+func TestBlobStoreBadger_NewIteratorAfterCloseDoesNotPanic(t *testing.T) {
+	// Create a Badger blob store.
+	store, err := New()
+	require.NoError(t, err)
+
+	// Create a transaction from that store.
+	txn := store.NewTransaction(false)
+
+	// Close the store.
+	require.NoError(t, store.Close())
+
+	// Try to create an iterator using the old transaction and expect no panic.
+	var iter types.BlobIterator
+	require.NotPanics(t, func() {
+		iter = store.NewIterator(txn, types.BlobIteratorOptions{
+			Prefix: []byte("x"),
+		})
+	})
+
+	// Expect an iterator object and to contain an error.
+	require.NotNil(t, iter)
+	require.Error(t, iter.Err())
 }

